File: integration_tests/java_client_tests_real/client-app/src/test/java/dev/graphlink/test/NestedTypesTest.java
package dev.graphlink.test;

import dev.graphlink.test.generated.client.GraphLinkClient;
import dev.graphlink.test.generated.enums.UserStatus;
import dev.graphlink.test.generated.types.*;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import static dev.graphlink.test.Fixtures.*;
import static org.junit.jupiter.api.Assertions.*;

class NestedTypesTest {

    private GraphLinkClient client;

    @BeforeEach
    void setUp() {
        client = newRealClient();
    }

    // ── Required nested object (User.address) ─────────────────────────────────

    @Nested
    class RequiredNestedObject {

        @Test
        void address_isNotNull() {
            assertNotNull(client.queries.getUser("user-1").getGetUser().getAddress());
        }

        @Test
        void address_hasCorrectStreet() {
            assertEquals("123 Main St",
                    client.queries.getUser("user-1").getGetUser().getAddress().getStreet());
        }

        @Test
        void address_hasCorrectCity() {
            assertEquals("Springfield",
                    client.queries.getUser("user-1").getGetUser().getAddress().getCity());
        }

        @Test
        void address_hasCorrectCountry() {
            assertEquals("US",
                    client.queries.getUser("user-1").getGetUser().getAddress().getCountry());
        }

        @Test
        void address_hasCorrectZip() {
            assertEquals("12345",
                    client.queries.getUser("user-1").getGetUser().getAddress().getZip());
        }

        @Test
        void address_deserializesToAddressType() {
            assertInstanceOf(Address.class,
                    client.queries.getUser("user-1").getGetUser().getAddress());
        }
    }

    // ── Post with required author (multi-level nesting) ───────────────────────

    @Nested
    class MultiLevelNesting {

        @Test
        void post_isNotNull() {
            assertNotNull(client.queries.getPost("post-1").getGetPost());
        }

        @Test
        void post_deserializesToPostType() {
            assertInstanceOf(Post.class, client.queries.getPost("post-1").getGetPost());
        }

        @Test
        void post_hasCorrectTitle() {
            assertEquals("Hello World",
                    client.queries.getPost("post-1").getGetPost().getTitle());
        }

        @Test
        void post_author_isNotNull() {
            assertNotNull(client.queries.getPost("post-1").getGetPost().getAuthor());
        }

        @Test
        void post_author_deserializesToUserFragmentType() {
            assertInstanceOf(User.class,
                    client.queries.getPost("post-1").getGetPost().getAuthor());
        }

        @Test
        void post_author_hasCorrectId() {
            assertEquals("user-1",
                    client.queries.getPost("post-1").getGetPost().getAuthor().getId());
        }

        @Test
        void post_author_status_deserializesToEnum() {
            assertEquals(UserStatus.ACTIVE,
                    client.queries.getPost("post-1").getGetPost().getAuthor().getStatus());
        }

        @Test
        void post_author_address_city_isAccessible() {
            assertEquals("Springfield",
                    client.queries.getPost("post-1").getGetPost().getAuthor().getAddress().getCity());
        }

        @Test
        void post_viewCount_deserializesToInteger() {
            assertEquals(128, client.queries.getPost("post-1").getGetPost().getViewCount());
        }
    }

    // ── @glSkipOnClient — server-only field silently ignored ──────────────────

    @Nested
    class SkipOnClientField {

        @Test
        void auditEntry_deserializesCorrectly() {
            GetAuditEntryResponse res = client.queries.getAuditEntry("audit-1");
            AuditEntry entry = res.getGetAuditEntry();
            assertNotNull(entry);
            assertEquals("audit-1", entry.getId());
            assertEquals("LOGIN", entry.getAction());
            assertEquals("user-1", entry.getPerformedBy());
        }

        @Test
        void auditEntry_withServerOnlyField_isIgnoredSilently() {
            GetAuditEntryResponse res = client.queries.getAuditEntry("audit-2");
            AuditEntry entry = res.getGetAuditEntry();
            assertNotNull(entry);
            assertEquals("audit-2", entry.getId());
            assertEquals("LOGOUT", entry.getAction());
        }
    }
}

File: integration_tests/swift_client_tests_real/Tests/GraphLinkGeneratedTests/NestedTypesTest.swift
import Foundation
import XCTest
@testable import GraphLinkGenerated

// ── Required nested object (User.address) ─────────────────────────────────

final class NestedTypesTest_RequiredNestedObject: XCTestCase {
    var client: GraphLinkClient!

    override func setUp() {
        client = Fixtures.newRealClient()
    }

    func test_address_isNotNull() async throws {
        let res = try await client.queries.getUser(id: "user-1")
        XCTAssertNotNil(res.getUser.address)
    }

    func test_address_hasCorrectStreet() async throws {
        let res = try await client.queries.getUser(id: "user-1")
        XCTAssertEqual(res.getUser.address.street, "123 Main St")
    }

    func test_address_hasCorrectCity() async throws {
        let res = try await client.queries.getUser(id: "user-1")
        XCTAssertEqual(res.getUser.address.city, "Springfield")
    }

    func test_address_hasCorrectCountry() async throws {
        let res = try await client.queries.getUser(id: "user-1")
        XCTAssertEqual(res.getUser.address.country, "US")
    }

    func test_address_hasCorrectZip() async throws {
        let res = try await client.queries.getUser(id: "user-1")
        XCTAssertEqual(res.getUser.address.zip, "12345")
    }

    func test_address_deserializesToAddressType() async throws {
        let res = try await client.queries.getUser(id: "user-1")
        XCTAssertTrue(res.getUser.address is Address)
    }
}

// ── Multi-level nesting (Post → User) ─────────────────────────────────────

final class NestedTypesTest_MultiLevelNesting: XCTestCase {
    var client: GraphLinkClient!

    override func setUp() {
        client = Fixtures.newRealClient()
    }

    func test_post_isNotNull() async throws {
        let res = try await client.queries.getPost(id: "post-1")
        XCTAssertNotNil(res.getPost)
    }

    func test_post_deserializesToPostType() async throws {
        let res = try await client.queries.getPost(id: "post-1")
        XCTAssertTrue(res.getPost is Post)
    }

    func test_post_hasCorrectTitle() async throws {
        let res = try await client.queries.getPost(id: "post-1")
        XCTAssertEqual(res.getPost.title, "Hello World")
    }

    func test_post_author_isNotNull() async throws {
        let res = try await client.queries.getPost(id: "post-1")
        XCTAssertNotNil(res.getPost.author)
    }

    func test_post_author_deserializesToFragmentType() async throws {
        let res = try await client.queries.getPost(id: "post-1")
        XCTAssertTrue(res.getPost.author is User)
    }

    func test_post_author_hasCorrectId() async throws {
        let res = try await client.queries.getPost(id: "post-1")
        XCTAssertEqual(res.getPost.author?.id, "user-1")
    }

    func test_post_author_status_deserializesToEnum() async throws {
        let res = try await client.queries.getPost(id: "post-1")
        XCTAssertEqual(res.getPost.author?.status, UserStatus.active)
    }

    func test_post_author_address_city_isAccessible() async throws {
        let res = try await client.queries.getPost(id: "post-1")
        XCTAssertEqual(res.getPost.author?.address.city, "Springfield")
    }

    func test_post_viewCount_deserializesToInt() async throws {
        let res = try await client.queries.getPost(id: "post-1")
        XCTAssertEqual(res.getPost.viewCount, 128)
    }
}

// ── @glSkipOnClient — server-only field silently ignored ──────────────────

final class NestedTypesTest_SkipOnClientField: XCTestCase {
    var client: GraphLinkClient!

    override func setUp() {
        client = Fixtures.newRealClient()
    }

    func test_auditEntry_deserializesCorrectly() async throws {
        let entry = try await client.queries.getAuditEntry(id: "audit-1").getAuditEntry
        XCTAssertNotNil(entry)
        XCTAssertEqual(entry.id, "audit-1")
        XCTAssertEqual(entry.action, "LOGIN")
        XCTAssertEqual(entry.performedBy, "user-1")
    }

    func test_auditEntry_withExtraServerField_isIgnoredSilently() async throws {
        let entry = try await client.queries.getAuditEntry(id: "audit-2").getAuditEntry
        XCTAssertNotNil(entry)
        XCTAssertEqual(entry.id, "audit-2")
        XCTAssertEqual(entry.action, "LOGOUT")
    }
}
